File: src/solution_java/hard/_23MergeKSortedLists.java
package solution_java.hard;

public class _23MergeKSortedLists {
    class ListNode {
        int val;
        ListNode next;

        public ListNode(int val) {
            this.val = val;
        }

        public void setNext(ListNode next) {
            this.next = next;
        }
    }

    public class Solution {
        public ListNode mergeKLists(ListNode[] lists) {
            return merge(lists, 0, lists.length - 1);
        }

        ListNode merge(ListNode[] lists, int left, int right) {
            if (left == right) {
                return lists[left];
            }
            if (left > right) {
                return null;
            }
            int middle = left + ((right - left) >> 1);
            return mergeTwoListNode(merge(lists, left, middle), merge(lists, middle + 1, right));
        }

        ListNode mergeTwoListNode(ListNode left, ListNode right) {
            if (left == null) {
                return right;
            }
            if (right == null) {
                return left;
            }
            ListNode head = new ListNode(0);
            ListNode preV = head;
            while (left != null && right != null) {
                if (left.val <= right.val) {
                    preV.next = left;
                    left = left.next;
                } else {
                    preV.next = right;
                    right = right.next;
                }
                preV = preV.next;
            }
            if (left == null) {
                preV.next = right;
            }
            if (right == null) {
                preV.next = left;
            }
            return head.next;
        }
    }
}

File: src/hard/_23MergeKSortedLists.kt
package hard

import ListNode

class _23MergeKSortedLists {
    class Solution {
        fun mergeKLists(lists: Array<ListNode?>): ListNode? {
            return merge(lists, 0, lists.size - 1)
        }

        private fun merge(lists: Array<ListNode?>, left: Int, right: Int): ListNode? {
            if (left == right) {
                return lists[left]
            }
            if (left > right) {
                return null
            }
            val middle = (left + right) shr 1
            return mergeTwoList(merge(lists, left, middle), merge(lists, middle + 1, right))
        }

        private fun mergeTwoList(tempList1: ListNode?, tempList2: ListNode?): ListNode? {
            if (tempList1 == null) {
                return tempList2
            }
            if (tempList2 == null) {
                return tempList1
            }
            val head = ListNode(0)
            var preV: ListNode
            var list1 = tempList1
            var list2 = tempList2
            if (list1.`val` <= list2.`val`) {
                preV = list1
                list1 = list1.next
            } else {
                preV = list2
                list2 = list2.next
            }
            head.next = preV

            while (list1 != null && list2 != null) {
                if (list1.`val` <= list2.`val`) {
                    preV.next = list1
                    list1 = list1.next
                } else {
                    preV.next = list2
                    list2 = list2.next
                }
                preV = preV.next!!
            }
            if (list1 == null) {
                preV.next = list2
            } else if (list2 == null) {
                preV.next = list1
            }

            return head.next
        }
    }
}